lib/s3_inventory.rb

```diff
@@ -0,0 +1,195 @@
+require "aws-sdk-s3"
+require "csv"
+require "discourse_event"
+
+class S3Inventory
+
+  class StorageError < RuntimeError; end
+
+  ::DiscourseEvent.on(:site_setting_saved) do |site_setting|
+    name = site_setting.name.to_s
+    Jobs.enqueue(:update_s3_inventory) if name.include?("s3_inventory") || name == "s3_upload_bucket"
+  end
+
+  attr_reader :inventory_id, :s3_client, :csv_filename,
+              :source_bucket_name, :source_bucket_path,
+              :destination_bucket_name, :destination_bucket_path
+
+  CSV_ETAG_INDEX ||= 2.freeze
+
+  def initialize(inventory_id = "uploads")
+    @source_bucket_name, @source_bucket_path = begin
+      raise Discourse::InvalidParameters.new("s3_upload_bucket") if SiteSetting.s3_upload_bucket.blank?
+      S3Helper.get_bucket_and_folder_path(SiteSetting.s3_upload_bucket)
+    end
+
+    @destination_bucket_name, @destination_bucket_path = begin
+      raise Discourse::InvalidParameters.new("s3_inventory_bucket") if SiteSetting.s3_inventory_bucket.blank?
+      S3Helper.get_bucket_and_folder_path(SiteSetting.s3_inventory_bucket)
+    end
+
+    s3_options = S3Helper.s3_options(SiteSetting)
+    @s3_helper = S3Helper.new(SiteSetting.s3_inventory_bucket, '', s3_options)
+    @s3_client = @s3_helper.s3_client
+
+    @inventory_id = inventory_id
+  end
+
+  def file
```

Comment:
Does this method need to be public?

Reply:
I need this for rspec test cases [here](https://github.com/discourse/discourse/pull/6867/files/6fa012dbf08131ac757ad2583a9f1c45ee402a14#diff-43df0e63549fdfd6a4400759c68ce74eR51).